raiden/network/proxies/token_network.py

```diff
@@ -812,96 +846,168 @@ def _set_total_deposit(
                 partner=partner,
             )
             self.client.poll(transaction_hash)
-            receipt_or_none = check_transaction_threw(self.client, transaction_hash)
+            receipt = check_transaction_threw(self.client, transaction_hash)
 
-        transaction_executed = gas_limit is not None
-        if not transaction_executed or receipt_or_none:
-            if transaction_executed:
-                block = receipt_or_none["blockNumber"]
-            else:
-                block = checking_block
+            if receipt:
+                # Because the gas estimation succeeded it is known that:
+                # - The channel was open.
+                # - The account had enough tokens to deposit
+                # - The account had enough balance to pay for the gas (however
+                #   there is a race condition for multiple transactions #3890)
+                failed_at_blockhash = encode_hex(receipt["blockHash"])
+
+                if receipt["cumulativeGasUsed"] == gas_limit:
+                    msg = (
+                        f"setTotalDeposit failed and all gas was used "
+                        f"({gas_limit}). Estimate gas may have underestimated "
+                        f"setTotalDeposit, or succeeded even though an assert is "
+                        f"triggered, or the smart contract code has an "
+                        f"conditional assert."
+                    )
+                    raise RaidenUnrecoverableError(msg)
+
+                # Query the current state to check for transaction races
```

Comment:
Perhaps something like:

s/current state/channel state when the transaction was mined/ 

can be better